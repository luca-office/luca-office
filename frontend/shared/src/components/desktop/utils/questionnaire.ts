import {ScenarioQuestionnairesQuery_scenarioQuestionnaires} from "../../../graphql/generated/ScenarioQuestionnairesQuery"

export const questionnaireTypeForQuestionId = (
  questionId: UUID,
  questionnaireId: UUID,
  scenarioQuestionnaires: ScenarioQuestionnairesQuery_scenarioQuestionnaires[]
) =>
  scenarioQuestionnaires
    .filter(scenarioQuestionnaire => scenarioQuestionnaire.questionnaireId === questionnaireId)
    .map(scenarioQuestionnaires => scenarioQuestionnaires.questionnaire)
    .flatMap(questionnaire => questionnaire.questions)
    .find(question => question.id === questionId)?.questionType
