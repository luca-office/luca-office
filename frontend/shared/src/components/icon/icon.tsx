import {css} from "@emotion/react"
import {kebabCase} from "lodash-es"
import * as React from "react"
import {IconName} from "../../enums"
import {CustomStyle, iconDefaultColor, spacingTiny} from "../../styles"

interface IconProps {
  readonly className?: string
  readonly color?: string
  readonly hoverColor?: string
  readonly customSpacing?: number
  readonly fillColor?: string
  readonly hasSpacing?: boolean
  readonly height?: number
  readonly onClick?: (event: React.MouseEvent<HTMLDivElement>) => void
  readonly title?: string
  readonly width?: number
}

interface Props extends IconProps, CustomStyle {
  readonly name: IconName
}

export const Icon: React.FC<Props> = props => (
  <div
    onClick={props.onClick}
    title={props.title}
    className={props.className}
    css={[iconStyle(props), props.customStyles]}
    dangerouslySetInnerHTML={{__html: content(props.name)}}
  />
)

// Dynamic import to avoid huge switch case here
const content = (iconName: IconName) => {
  const name = kebabCase(iconName)
  return require(`../../assets/images/icons/${name}.svg`)
}

Icon.defaultProps = {
  width: 16,
  height: 16
}

const iconStyle = (props: Props) =>
  css({
    height: props.height,
    width: props.width,
    minWidth: props.width,
    color: props.color ?? iconDefaultColor,
    ...(props.hasSpacing ? {marginRight: props.customSpacing || spacingTiny} : {}),

    ...(!!props.hoverColor && {
      "&:hover": {
        color: props.hoverColor
      }
    })
  })
