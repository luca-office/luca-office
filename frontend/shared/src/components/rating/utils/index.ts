export * from "./coding-criterion"
export * from "./coding-dimension"
export * from "./coding-item"
export * from "./question"
export * from "./score"
export * from "./scoring-type"
