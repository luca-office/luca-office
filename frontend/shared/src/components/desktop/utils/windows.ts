import findLastIndex from "lodash-es/findLastIndex"
import {OfficeTool, OfficeWindowType} from "../../../enums"
import {Option} from "../../../utils/option"

export const findVisibleWindows = (openedWindows: OfficeWindowType[], minimizedWindows: OfficeWindowType[]) => {
  const visibleWindows = openedWindows.filter(window => !minimizedWindows.includes(window))

  return Option.of(findLastIndex(visibleWindows, isFullScreenWindow))
    .map(lastFullScreenToolIndex => visibleWindows.slice(lastFullScreenToolIndex))
    .getOrElse(visibleWindows)
}

const isFullScreenWindow = (window: OfficeWindowType) => {
  switch (window) {
    case OfficeTool.Notes:
    case OfficeTool.Calculator:
    case OfficeTool.Chat:
      return false
    default:
      return true
  }
}
