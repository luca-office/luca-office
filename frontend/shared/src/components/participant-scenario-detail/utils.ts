import {FreetextAnswerFragment} from "../../graphql/generated/FreetextAnswerFragment"
import {QuestionType} from "../../graphql/generated/globalTypes"
import {Questionnaire, RuntimeSurveyResults} from "../../models"
import {ParticipantQuestionAnswer, QuestionnaireParticipantResults} from "../questionnaire"

export const runtimeSurveyResultsToQuestionnaireParticipantResults = (
  invitationId: UUID,
  questionnaire: Questionnaire,
  runtimeSurveyResults: RuntimeSurveyResults,
  freeTextAnswers: FreetextAnswerFragment[]
): QuestionnaireParticipantResults => {
  const answers = runtimeSurveyResults.questionResults.reduce<Record<UUID, ParticipantQuestionAnswer>>(
    (answersAccumulator, result) => {
      const question = questionnaire.questions.find(question => question.id === result.questionId)
      const participantResult = result.participantResults.find(results => results.invitationId === invitationId)
      const freeTextAnswer = freeTextAnswers.find(
        answer => answer.questionId === result.questionId && answer.surveyInvitationId === invitationId
      )?.text

      switch (question?.questionType) {
        case QuestionType.SingleChoice: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.SingleChoice,
              questionId: question.id,
              selectedAnswerId: participantResult?.selectedAnswerIds[0] ?? "",
              isFreetextAnswerSelected: participantResult?.wasFreetextAnswerSelected ?? false,
              freeTextAnswer: freeTextAnswer
            }
          }
        }

        case QuestionType.MultipleChoice: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.MultipleChoice,
              questionId: question.id,
              selectedAnswerIds: participantResult?.selectedAnswerIds ?? [],
              isFreetextAnswerSelected: participantResult?.wasFreetextAnswerSelected ?? false,
              freeTextAnswer: freeTextAnswer
            }
          }
        }

        case QuestionType.FreeText: {
          return {
            ...answersAccumulator,
            [question.id]: {
              questionType: QuestionType.FreeText,
              questionId: question.id,
              freeTextAnswer: freeTextAnswer ?? ""
            }
          }
        }

        default:
          return answersAccumulator
      }
    },
    {}
  )

  return {questionnaireId: runtimeSurveyResults.questionnaireId, answers}
}
