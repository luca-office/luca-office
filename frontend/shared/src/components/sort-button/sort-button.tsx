import {css} from "@emotion/react"
import * as React from "react"
import {IconName} from "../../enums"
import {CustomStyle} from "../../styles"
import {Option} from "../../utils"
import {Icon, Tooltip} from ".."
import {TableSortState} from "../table/sort-utils"

interface Props extends CustomStyle {
  readonly tableSortState: Option<TableSortState>
  readonly onClick: () => void
  readonly toolTipText?: string
}

export const SortButton: React.FC<Props> = ({onClick, tableSortState, toolTipText, customStyles}) => (
  <Tooltip inactive={toolTipText === undefined} title={toolTipText ?? ""}>
    <Icon
      customStyles={[styles.icon, customStyles]}
      onClick={onClick}
      name={tableSortState.map(sortState => iconForSortState(sortState)).getOrElse(IconName.TriangleRight)}
    />
  </Tooltip>
)

const iconForSortState = (sortState: TableSortState): IconName => {
  switch (sortState) {
    case TableSortState.ASCENDING:
      return IconName.TriangleDown
    case TableSortState.DESCENDING:
      return IconName.TriangleUp
    default:
      return IconName.TriangleRight
  }
}

const styles = {
  icon: css({
    cursor: "pointer"
  })
}
