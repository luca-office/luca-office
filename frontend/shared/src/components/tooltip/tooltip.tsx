import * as React from "react"
import {css} from "@emotion/react"
import {CSSInterpolation} from "@emotion/serialize"
import {Tooltip as ReactTooltip} from "react-tooltip"
import {HeadingLevel, IconName} from "../../enums"
import {
  boxSizeLarge,
  Children,
  CustomStyle,
  Flex,
  fontColor,
  fontColorDark,
  fontColorLight,
  fontFamily,
  spacing,
  spacingSmall,
  spacingTinier
} from "../../styles"
import {createUUID} from "../../utils/uuid"
import {Icon} from "../icon/icon"
import {ReactPortal} from "../react-portal/react-portal"
import {Heading, Text} from "../typography/typography"

export interface TooltipHintProps extends Children {
  readonly title: string
  readonly delayShow?: number
  readonly text?: string
  readonly icon?: IconName
  readonly iconColor?: string
  readonly renderCustomContent?: () => JSX.Element
  readonly backgroundColor?: string
  readonly placement?: "top" | "right" | "bottom" | "left" | undefined
  readonly inactive?: boolean
  readonly withPortal?: boolean
  readonly onWrapperClick?: () => void
  readonly customTooltipStyles?: CSSInterpolation
}

export const Tooltip: React.FC<TooltipHintProps & CustomStyle> = ({
  children,
  title,
  text,
  icon,
  delayShow,
  iconColor,
  placement = "bottom",
  inactive,
  backgroundColor,
  renderCustomContent,
  customStyles,
  withPortal,
  onWrapperClick,
  customTooltipStyles
}) => {
  const uuid = `tooltip_${createUUID()}`

  if (inactive) {
    return <React.Fragment>{children}</React.Fragment>
  }

  const renderTooltipContent = () => (
    <div css={[contentStyles.wrapper, icon && contentStyles.wrapperWithIcon]}>
      <div css={contentStyles.textWrapper}>
        <Heading level={HeadingLevel.h3} customStyles={contentStyles.text}>
          {title}
        </Heading>
        <Text customStyles={contentStyles.text}>{text}</Text>
      </div>
      {icon && (
        <div css={contentStyles.iconWrapper}>
          <Icon name={icon} color={iconColor ?? fontColorLight} width={Sizes.iconSize} height={Sizes.iconSize} />
        </div>
      )}
    </div>
  )

  const tooltip = (
    <ReactTooltip
      anchorId={uuid}
      place={placement}
      delayShow={delayShow}
      style={{zIndex: 1, backgroundColor: backgroundColor ?? fontColor}}
      css={[styles.tooltip, customTooltipStyles]}
      content={title}
      children={text ? renderTooltipContent() : renderCustomContent ? renderCustomContent() : undefined}
    />
  )

  return (
    <React.Fragment>
      <div css={customStyles} id={uuid} onClick={onWrapperClick}>
        {children}
      </div>

      {withPortal ? <ReactPortal>{tooltip}</ReactPortal> : tooltip}
    </React.Fragment>
  )
}

// !important needed to override tooltip default padding to fit requirement
const styles = {
  tooltip: css`
    padding: 2px 8px !important;
    font-family: ${fontFamily};
  `
}

const Sizes = {
  tipSize: 288,
  iconSize: 48
}

const contentStyles = {
  wrapper: css({
    display: "grid",
    margin: spacing(-spacingTinier, -spacingSmall),
    width: Sizes.tipSize
  }),
  wrapperWithIcon: css({
    gridTemplateColumns: `1fr ${boxSizeLarge}px`
  }),
  textWrapper: css({
    // bottom padding accounts for text spacer
    padding: spacing(spacingSmall, spacingSmall, spacingSmall - spacingTinier),
    textAlign: "left",
    boxSizing: "border-box",
    overflow: "hidden"
  }),
  text: css({
    color: "white",
    marginBottom: spacingTinier,
    whiteSpace: "initial"
  }),
  iconWrapper: css(Flex.row, {
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: fontColorDark
  })
}
