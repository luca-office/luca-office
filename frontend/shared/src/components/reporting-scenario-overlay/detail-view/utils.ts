import {TableEntity} from "../../../components/rating/models"
import {CodingDimension, CodingItem, CodingItemResultByItemId} from "../../../models"

export const itemToTableEntity = (
  item: CodingItem,
  codingItemResultsByItem: CodingItemResultByItemId
): TableEntity => ({
  id: item.id,
  title: item.title,
  position: item.position,
  score: codingItemResultsByItem[item.id].score,
  maxScore: codingItemResultsByItem[item.id].maximumScore,
  averageScore: codingItemResultsByItem[item.id].averageScore,
  rated: true
})

export const dimensionToTableEntity = (
  dimension: CodingDimension,
  items: CodingItem[],
  codingItemResultsByItem: CodingItemResultByItemId
): TableEntity => ({
  id: dimension.id,
  title: dimension.title,
  position: dimension.position,
  score: items.reduce((acc, item) => acc + codingItemResultsByItem[item.id].score, 0),
  maxScore: items.reduce((acc, item) => acc + codingItemResultsByItem[item.id].maximumScore, 0),
  averageScore: items.reduce((acc, item) => acc + codingItemResultsByItem[item.id].averageScore, 0),
  rated: true
})
