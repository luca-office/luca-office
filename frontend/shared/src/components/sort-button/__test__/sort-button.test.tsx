// importing from direct file because of issues of babel loader and spyOn
import {shallow} from "enzyme"
import * as React from "react"
import {create} from "react-test-renderer"
import {Option} from "../../../utils"
import {Icon} from "../../icon/icon"
import {Tooltip} from "../../tooltip/tooltip"
import {SortButton} from "../sort-button"

const sortButton = <SortButton onClick={jest.fn()} tableSortState={Option.none()} toolTipText="text" />

describe("sort-button", () => {
  it("renders correctly", async () => {
    const component = create(sortButton)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
  it("has correct structure", async () => {
    const component = shallow(sortButton)

    expect(component.find(Icon)).toHaveLength(1)
    expect(component.find(Tooltip)).toHaveLength(1)
  })
  it("triggers action correctly", async () => {
    const handleClick = jest.fn()
    const component = shallow(<SortButton onClick={handleClick} tableSortState={Option.none()} toolTipText="text" />)

    component.find(Icon).simulate("click")
    expect(handleClick).toBeCalledTimes(1)
  })
})
