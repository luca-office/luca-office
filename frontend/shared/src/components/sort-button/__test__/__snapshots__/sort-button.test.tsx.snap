// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sort-button renders correctly 1`] = `
[
  <div
    className="css-0"
    id="tooltip_42"
  >
    <div
      className="css-ivyasj-Icon"
      dangerouslySetInnerHTML={
        {
          "__html": {},
        }
      }
      onClick={[MockFunction]}
    />
  </div>,
  <div
    className="react-tooltip styles-module_tooltip__mnnfp styles-module_dark__xNqje css-10zyuev-Tooltip"
    role="tooltip"
    style={
      {
        "backgroundColor": "#2e3032",
        "zIndex": 1,
      }
    }
  >
    text
    <div
      className="react-tooltip-arrow styles-module_arrow__K0L3T"
      style={{}}
    />
  </div>,
]
`;
